File: native/ApplicationInsightsObjectiveC/ApplicationInsightsObjectiveC/MSAIData.m
#import "MSAIData.h"
#import "MSAIOrderedDictionary.h"

/// Data contract class for type Data.
@implementation MSAIData

///
/// Adds all members of this class to a dictionary
/// @param dictionary to which the members of this class will be added.
///
- (MSAIOrderedDictionary *)serializeToDictionary {
    MSAIOrderedDictionary *dict = [super serializeToDictionary];
    MSAIOrderedDictionary *baseDataDict = [self.baseData serializeToDictionary];
    if ([NSJSONSerialization isValidJSONObject:baseDataDict]) {
        [dict setObject:baseDataDict forKey:@"baseData"];
    } else {
        NSLog(@"[ApplicationInsights] Some of the telemetry data was not NSJSONSerialization compatible and could not be serialized!");
    }
    return dict;
}

#pragma mark - NSCoding

- (instancetype)initWithCoder:(NSCoder *)coder {
  self = [super initWithCoder:coder];
  if(self) {
    _baseData = [coder decodeObjectForKey:@"self.baseData"];
  }

  return self;
}

- (void)encodeWithCoder:(NSCoder *)coder {
  [super encodeWithCoder:coder];
  [coder encodeObject:self.baseData forKey:@"self.baseData"];
}


@end

File: native/SwiftInsights/SwiftInsights/MSAIData.swift
import Foundation

class MSAIData: NSObject, NSCoding {
    
    var baseData: MSAIBaseData?
    
    override init() {
        super.init()
    }
    
    func serializeToDictionary() -> [String: Any] {
        var dict = [String: Any]()
        if let baseDataDict = baseData?.serializeToDictionary(),
           JSONSerialization.isValidJSONObject(baseDataDict) {
            dict["baseData"] = baseDataDict
        } else {
            print("[ApplicationInsights] Some of the telemetry data was not NSJSONSerialization compatible and could not be serialized!")
        }
        return dict
    }
    
    // MARK: - NSCoding
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        self.baseData = coder.decodeObject(forKey: "self.baseData") as? MSAIBaseData
    }
    
    func encode(with coder: NSCoder) {
        super.encode(with: coder)
        coder.encode(baseData, forKey: "self.baseData")
    }
}
